lib/Driver/ToolChains.cpp

```diff
@@ -1190,11 +1190,52 @@ getSectionMagicArch(const llvm::Triple &Triple) {
   }
 }
 
+std::string toolchains::GenericUnix::getDefaultLinker() const {
+  switch(getTriple().getArch()) {
+  case llvm::Triple::arm:
+  case llvm::Triple::armeb:
+  case llvm::Triple::thumb:
+  case llvm::Triple::thumbeb:
+    // BFD linker has issues wrt relocation of the protocol conformance
+    // section on these targets, it also generates COPY relocations for
+    // final executables, as such, unless specified, we default to gold
+    // linker.
+    return "gold";
+  default:
+    // Otherwise, use the default BFD linker.
+    return "";
+  }
+}
+
+bool toolchains::GenericUnix::shouldProvideRPathToLinker() const {
+  return true;
+}
+
+bool toolchains::GenericUnix::shouldSpecifyTargetTripleToLinker() const {
+  return true;
+}
+
+std::string toolchains::GenericUnix::getPreInputObjectPath(
+       const SmallString<128> RuntimeLibraryPath) const {
```

Comment:
C++ note: please use `StringRef` here (no `const` needed), so that we reference the data in the existing SmallString. This is an implicit conversion in LLVM, so no need to change the call site.